File: frontend/src/api/organisations.ts
import { api } from ".";
import type { UserProfile } from "./profiles";

// Models

export type Organisation = {
    id: string;
    name: string;
    slug: string;
    logo_url: string;
    created_by: string; // uuid (owner)
    created_at: string; // ISO string
};

export type OrganisationRole = "owner" | "admin" | "member";

export type OrganisationMember = {
    user: UserProfile;
    role: OrganisationRole; // owner / admin / member
    joined_at: string; // ISO string
};

export type CreateOrganisationRequest = {
    name: string;
    logo_url?: string;
};

// API Calls

// Fetches all organisations of the signed in user
export const fetchOrganisations = async () => {
    const res = await api.get<Organisation[]>("/v1/organisations");
    return res.data;
};

// Fetches all members of an organisations
export const fetchOrganisationMembers = async (org_id: string) => {
    const res = await api.get<OrganisationMember[]>(
        "/v1/organisations/" + org_id + "/members"
    );
    return res.data;
};

// Creates a new organisation with the signed in user as the owner
// Returns the new organisation
export const createOrganisation = async (body: CreateOrganisationRequest) => {
    const res = await api.post<Organisation>("/v1/organisations", body);
    return res.data;
};

File: backend/internal/models/organisations.go
package models

import "time"

// Database Models

type Organisation struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Slug      string    `json:"slug"`
	LogoURL   string    `json:"logo_url"`
	CreatedBy string    `json:"created_by"` // organisation owner
	CreatedAt time.Time `json:"created_at"`
}

type OrganisationRole string

const (
	RoleOwner  OrganisationRole = "owner"
	RoleAdmin  OrganisationRole = "admin"
	RoleMember OrganisationRole = "member"
)

type OrganisationMember struct {
	OrganisationID string           `json:"organisation_id"`
	UserID         string           `json:"user_id"`
	Role           OrganisationRole `json:"role"`       // owner / admin / member
	InvitedBy      string           `json:"invited_by"` // user id
	CreatedAt      time.Time        `json:"created_at"`
}

// Data Transfer Objects (for sending data to frontend)

type OrganisationMemberDTO struct {
	User     UserProfile `json:"user"`
	Role     string      `json:"role"`
	JoinedAt time.Time   `json:"joined_at"` // created_at within db
}
